### PR TITLE
TEP-0118: matrix: add `matrix.params` field

### PR DESCRIPTION
**Pull request by chengjoey**
# Changes
[TEP-0118](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md)  proposed move the existing array of Parameters to a field - `params` - within the matrix field

In this change, we move the existing field - `params` - within the matrix

# API changes 
1. move the existing array of Parameters to a field - `params` - within the matrix : params

# TODO
add `include` filed within the matrix and implement  `include` specific combinations in the Matrix

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: matrixed params must be moved under `matrix.params`
```

### CODE REVIEW

**Review by jerop — COMMENTED** [review]
This is not just a refactor because it's changing the API by adding a `matrix.params` field -- please update the commit message, PR description, release notes and documentation to explain the new field

**Comment by chengjoey** [comment]
> This is not just a refactor because it's changing the API by adding a `matrix.params` field -- please update the commit message, PR description, release notes and documentation to explain the new field

hi, I have adjusted my PR description replease notes and documentation as suggested, please do some reviewing and give some advice

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 1]
```diff
@@ -220,6 +219,28 @@ type PipelineTask struct {
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 }
 
+// Matrix is used to fan out Tasks in a Pipeline
+type Matrix struct {
+	// Params is a list of parameters used to fan out the pipelineTask
+	// Params takes only `Parameters` of type `"array"`
+	// Each array element is supplied to the `PipelineTask` by substituting `params` of type `"string"` in the underlying `Task`.
+	// The names of the `params` in the `Matrix` must match the names of the `params` in the underlying `Task` that they will be substituting.
+	Params []Param `json:"params,omitempty"`
```
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:228]: you might need a `// +listType=atomic` annotation for these new fields to resolve the build test failures.

File: pkg/apis/pipeline/v1beta1/pipeline_types.go [hunk 2]
```diff
@@ -220,6 +219,28 @@ type PipelineTask struct {
...
+
+	// Include is a list of MatrixInclude
+	Include []Include `json:"include,omitempty"`
+}
+
+// Include is used to specify combinations in the matrix
+type Include struct {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipeline_types.go:235]: It looks like `include` is introduced in this PR but not used anywhere. I'd suggest keeping the scope of this PR to just move existing matrix params under the `matrix.params` field, and introduce `include` in a later PR (and update the commit message/pr description and title accordingly)

**Comment by lbernick** [comment]
@chengjoey just updated/clarified your release note to include "action required"-- this string is parsed by our release notes script so it can be more easily highlighted to users

File: pkg/apis/pipeline/v1beta1/pipeline_types.go (hunk 1 above)
Comment (chengjoey) [reply pkg/apis/pipeline/v1beta1/pipeline_types.go]: thanks @lbernick ,I've added the annotation `// +listType=atomic` for `params` filed

File: pkg/apis/pipeline/v1beta1/pipeline_types.go (hunk 2 above)
Comment (chengjoey) [reply pkg/apis/pipeline/v1beta1/pipeline_types.go]: ok，I've removed the `include` and just move existing matrix params under the matrix.params field.

commit message/pr description and title are updated

**Comment by lbernick** [comment]
Thanks for making those changes! I'm sorry, I forgot to mention yesterday, would you mind also making these changes in the [v1 package](https://github.com/tektoncd/pipeline/blob/a1ca389bddfce53f0d5b6990f841a1f6617db520/pkg/apis/pipeline/v1/pipeline_types.go#L193)

**Comment by chengjoey** [comment]
> Thanks for making those changes! I'm sorry, I forgot to mention yesterday, would you mind also making these changes in the [v1 package](https://github.com/tektoncd/pipeline/blob/a1ca389bddfce53f0d5b6990f841a1f6617db520/pkg/apis/pipeline/v1/pipeline_types.go#L193)

ok, I'll also make these changes in v1

**Comment by chengjoey** [comment]
> Thanks for making those changes! I'm sorry, I forgot to mention yesterday, would you mind also making these changes in the [v1 package](https://github.com/tektoncd/pipeline/blob/a1ca389bddfce53f0d5b6990f841a1f6617db520/pkg/apis/pipeline/v1/pipeline_types.go#L193)

hi, I've make these changes in v1 package

**Review by lbernick — APPROVED** [review]
thank you! Please squash into one commit when you have the chance

**Comment by chengjoey** [comment]
> thank you! Please squash into one commit when you have the chance

Thank you for your memtion, i've squashed commits into one

File: pkg/apis/pipeline/v1/pipeline_types.go [hunk 3]
```diff
@@ -264,8 +273,13 @@ func (pt PipelineTask) validateTask(ctx context.Context) (errs *apis.FieldError)
 	return errs
 }
 
+// IsMatrixed return whether pipeline task is matrixed
+func (pt *PipelineTask) IsMatrixed() bool {
+	return pt.Matrix != nil && len(pt.Matrix.Params) > 0
+}
+
```
Comment (jerop) [inline pkg/apis/pipeline/v1/pipeline_types.go:280]: please add tests for this new exported function

File: pkg/apis/pipeline/v1beta1/pipeline_types_test.go [hunk 4]
```diff
@@ -1057,3 +1074,49 @@ func TestPipelineTask_ValidateEmbeddedOrType(t *testing.T) {
...
+			pt := &PipelineTask{
+				Matrix: tc.arg.Matrix,
+			}
+			isMatrixed := pt.IsMatrixed()
+			if isMatrixed != tc.expected {
+				t.Errorf("PipelineTask.IsMatrixed() return bool: %v, but wanted: %v", isMatrixed, tc.expected)
+			}
+		})
+	}
+}
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/pipeline_types_test.go:1122]: this test needs to be added in v1 too

Comment (chengjoey) [reply pkg/apis/pipeline/v1beta1/pipeline_types_test.go:1122]: thanks @jerop , i've added this test to v1

**Review by jerop — APPROVED** [review]
@chengjoey thank you and welcome to Tekton! 

/lgtm
/meow

@chengjoey would love to learn more about your interest in Matrix and ask if you have any early feedback from using it - couldn't find you on slack, please message me if you'd like to chat 😀

**Comment by chengjoey** [comment]
> would be mero than happy discuss the matrix with you, I have messaged you on slack,but i'm not sure if it's you😀

**Comment by chengjoey** [comment]
> @chengjoey thank you and welcome to Tekton!
> 
> /lgtm /meow
> 
> @chengjoey would love to learn more about your interest in Matrix and ask if you have any early feedback from using it - couldn't find you on slack, please message me if you'd like to chat 😀

would be mero than happy discuss the matrix with you, I have messaged you on slack,but i'm not sure if it's you😀

**Comment by chengjoey** [comment]
> New changes are detected. LGTM label has been removed.

rebase and solve the conflict files

**Pull request merged**